### PR TITLE
Adds ValidateUserDelete to validation webhook

### LINKED ISSUE #11275: Add ValidationDelete for Users Validation Webhook
### Description of the feature you would like to add / User story

As a developer
I would like to add ValidationDelete
in order to not allow deletion of users if the user is a sole owner of the project or has clusters/external clusters existing

### Solution details

- Requirement 1
- Requirement 2
- Requirement 3

### PR DESCRIPTION
**Pull request by imharshita**
Signed-off-by: Harshita sharma <harshita.sharma6174@gmail.com>

**What this PR does / why we need it**: Adds ValidateUserDelete to validation webhook
user deletion will not be allowed if the user is the sole owner of a project which has resources: clusters/external clusters

**Which issue(s) this PR fixes**:

Fixes #11275

**What type of PR is this?**

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
User deletion will not be allowed if the user is the sole owner of project which has resources i.e, clusters or external clusters
```

**Documentation**:

```documentation
NONE
```

### CODE REVIEW

File: pkg/apis/kubermatic/v1/helper/helper.go [hunk 1]
```diff
@@ -353,3 +354,37 @@ func UpdateResourceQuotaStatus(ctx context.Context, client ctrlruntimeclient.Cli
...
+		return false, fmt.Errorf("failed to list clusters: %w", err)
+	}
+
+	for _, cluster := range clusterList.Items {
+		if val, ok := cluster.ObjectMeta.Labels[kubermaticv1.ProjectIDLabelKey]; ok {
+			if val == projectName {
+				return true, nil
+			}
+		}
+	}
```
Comment (xrstf) [inline pkg/apis/kubermatic/v1/helper/helper.go]: You can just use a label selector.

File: pkg/validation/user.go [hunk 2]
```diff
@@ -85,3 +87,32 @@ func ValidateUserUpdate(oldUser, newUser *kubermaticv1.User) field.ErrorList {
...
+		// if project has multiple owner users
+		if len(project.OwnerReferences) > 1 {
+			continue
+		}
+
+		// project has single owner user then check if project has resources
+		hasClusters, err := kubermaticv1helper.HasClusters(ctx, client, project.Name)
+		if err != nil {
+			return err
+		}
```
Comment (xrstf) [inline pkg/validation/user.go]: This is not possible, Cluster objects live on seed clusters, users live on the master.

Comment (imharshita) [reply pkg/validation/user.go]: oh right!
I can try to get seed client then to fetch the clusters

File: pkg/validation/user.go [hunk 3]
```diff
@@ -85,3 +88,56 @@ func ValidateUserUpdate(oldUser, newUser *kubermaticv1.User) field.ErrorList {
...
+
+		// if project has clusters on any seed
+		listOpts := &ctrlruntimeclient.ListOptions{
+			Namespace: namespace,
+		}
+		seeds := &kubermaticv1.SeedList{}
+		if err := client.List(ctx, seeds, listOpts); err != nil {
+			return fmt.Errorf("failed to list the seeds: %w", err)
+		}
+		for _, seed := range seeds.Items {
```
Comment (xrstf) [inline pkg/validation/user.go]: Please use a seedsGetter instead. Same place where you got your seedClientGetter from.

File: pkg/apis/kubermatic/v1/helper/helper.go [hunk 4]
```diff
@@ -353,3 +354,21 @@ func UpdateResourceQuotaStatus(ctx context.Context, client ctrlruntimeclient.Cli
 		return client.Status().Patch(ctx, resourceQuota, ctrlruntimeclient.MergeFrom(original))
 	})
 }
+
+func HasClusters(ctx context.Context, client ctrlruntimeclient.Client, projectID string) (bool, error) {
+	clusterList := &kubermaticv1.ClusterList{}
+	if err := client.List(ctx, clusterList, ctrlruntimeclient.MatchingLabels{kubermaticv1.ProjectIDLabelKey: projectID}); err != nil {
```
Comment (xrstf) [inline pkg/apis/kubermatic/v1/helper/helper.go]: You should be able to limit this list to max. 1 item, because fetching more than 1 is wasted effort.

**Review by xrstf — APPROVED** [review]
/approve

**Comment by xrstf** [comment]
I am worried what happens when a single seed goes down, bbut if product wants to have it...

**Pull request merged**